pkg/mutation/system.go

```diff
@@ -148,20 +148,6 @@ func (s *System) Mutate(obj *unstructured.Unstructured, ns *corev1.Namespace) (b
 		}
 
 		if cmp.Equal(old, obj) {
-			if i == 0 {
-				convergence = SystemConvergenceTrue
-				return false, nil
-			}
-			if cmp.Equal(original, obj) {
```

Comment:
why do we not need these two checks anymore?

Reply:
I should not have deleted the first if-statement, since that is returning whether there were any mutations at all, which is valuable. Thanks for highlighting this, re-introducing.

The second check (`if cmp.Equal(original, obj)`) spends CPU cycles, but should never return `true` barring some truly weird bug. It is attempting to detect a cycle by checking whether the final value == the original value but:

- If there is a cycle, it's not guaranteed to be at the start of the iteration (see those programmer interview questions about finding loops in linked lists)
- This test only occurs after `obj == previousObj` (where `previousObj` is the object pre-mutation). In order for that to happen after the first iteration (when that is definitively true true b/c nothing was mutated), the outcome of applying a mutation to an object needs to be non-deterministic (not just non-idempotent). We can tell it requires non-determinism because, in order for this test to be reached, the first application of mutations to `original` must have yielded a change (otherwise the loop would have exited), but the second application to `original` must not yield a change. This implies this test would only detect randomly occurring bugs and only very selectively (only if the first object is matched).

The check is harmless from a logic perspective, but because deep comparisons are relatively expensive and the code is distracting from how the system actually behaves and is unnecessary complexity, IMO it's best to remove it.